### DISCUSSION TITLE
How to make a custom trigger (condition) in my acap application? Like in Axis Video Motion Detector

### DISCUSSION THREAD

**Discussion by AlexeyKozlovsky** [AxisCommunications/discussions#62] [ACAP] [answered]
The problem is that I want to make a custom trigger (or condition) which a user will be able to use in order to create events which they want to use. For example, I have a detector (acap application), if it's alarmed, the event which the user has created should be triggered. And, as I see, it may be possible for this event to be any kind. I just want to trigger it in the acap application side, and then, for example, event sends a snapshot via http or maybe create a text overlay, it doesn't matter what it will do. That's why it's important that a user should have an opportunity to set up this event (or not).
I used boost earlier and sent a request on my acap application side in a background thread. But if the problem which I described above can be solved, I would really appreciate your help. If there is no solution, I would also be glad to see your reply as it will save my time.

My application is in c++, I used acap-3 sdk (3.5). My device is axis M1135 on 10.11.65 firmware.

P.s. I saw in the acap documentation, that some Axis applications produce their custom triggers, that's why I think that it may be possible and for other developers. I'm attaching screenshot from video from this docs. There's a group of conditions "Applications" which seems to be "custom".

P.p.s. I thought that I suppose to use axevent, but it can be helpful if I want send or subscribe specific event, while I need only condition.

Thank you in advanced!
![image](https://user-images.githubusercontent.com/63499468/191284024-72844c66-b297-413e-b561-f2521f39d14b.png)

**Comment by mattias-kindborg-at-work** [comment]
Hi @AlexeyKozlovsky. We have an ACAP version 3 example [here](https://github.com/AxisCommunications/acap3-examples/tree/main/axevent). Please take a look and see if it's applicable to your problem.

**Reply by shadysp** [reply, marked as answer]
First one  need to use the right name spae.
The namespace tns1 is defined by Onvif and tnxaxis is defined by Axis.

About the events name spaces and topics please see that link.
https://www.axis.com/techsup/developer_doc/acap3/3.5/api/axevent/html/axevent_8h.html

From that link, here are some explanation :
All events are expressed as a set of key/value pairs, where the key should be assigned to a namespace. For example, the event describing that the first I/O port is active is described by the following key/value pairs.
   tns1:topic0=Device
tnsaxis:topic1=IO
tnsaxis:topic2=Port
          port=0
        state=1
Where
tns1: is a name space.
topic0: is a key.
Device: is a value.
 
So in this example you have event produced by the device/camera, the camera itself have different events that could be triggered upon, like temperature, storage, tampering, and different device status.
 
tns1:topic0=Device     (Declare that this is a camera event, not an application CameraApplicationPlatform)
 
tnsaxis:topic1=IO        (It is the input and output source, not MyApplicationName )
 
tnsaxis:topic2=Port      (The I/O source is a  port
         port=0                            first port
          state=1                            Active state)
 
Following that example you can declare the events in similar manner and then it would be possible to trigger and choose condition similar to other applications .
 
tns1:topic0=CameraApplication

tnsaxis:topic1=MyApplicationName
 
tnsaxis:topic2=Events
          event=alarm-1
         state=1 or 0
 
tnsaxis:topic2=Events
          event=alarm-2
         state=0 or 1
 
tnsaxis:topic2=Events
          event=alarm-3
         state=0 or 1
 
 
·         One could specify to a single event:
 
      tns1:topic0=CameraApplication
       tnsaxis:topic1=MyApplicationName
        tnsaxis:topic2=Events
         event=alarm-1
          state=1 or 0
 
·         Or All
 
       tns1:topic0=CameraApplication
        tnsaxis:topic1=MyApplicationName
         tnsaxis:topic2=Events
